File: semestr1/homework8/dz8_task1/HashTable.cpp
#include <iostream>
#include "HashTable.h"
#include "List.h"

const int hashSize = 100;

struct HashTable
{
	List *buckets[hashSize];
};

HashTable *createHashTable()
{
	HashTable *result = new HashTable;

	for (int i = 0; i < hashSize; i++)
		result->buckets[i] = createList();

	return result;
}

int sizeHashTable(HashTable *hash)
{
	return hashSize;
}

int hashFunction(char *value)
{
	int result = 0;

	for (int i = 0; value[i] != '\0'; i++)
	{
		result = (result + value[i]) % hashSize;
	}
	
	return result;
}

void addToHash(HashTable *hash, char *string)
{
	int index = hashFunction(string);
	bool successfulySearch = exists(hash->buckets[index], string);

	if (!successfulySearch)
		addToEnd(hash->buckets[index], string);
}

void deleteHashTable(HashTable *hash)
{
	for (int i = 0; i < hashSize; i++)
		deleteList(hash->buckets[i]);

	delete hash;
}

void printHash(HashTable *hash)
{
	for (int i = 0; i < hashSize; i++)
	{
		List *list = hash->buckets[i];
		ListElement *position = firstElement(list);
		int sizeList = returnAmountList(list);

		for (int j = 0; j < sizeList; j++)
		{
			std::cout << returnValuePosition(list, position) << ' ' << returnAmountElement(list, position) << std::endl;;
			position = nextElement(list, position);
		}
	}
}

File: semestr2/homework3/hw3_task4/src/hashTable/HashTable.java
package hashTable;

/**
 * use to create a hash table
 *
 * @author vladimir-zakharov
 */
public class HashTable {

    /**
     * constructor for hash table
     */
    public HashTable(HashFunctionInterface hashFunction) {
        for (int i = 0; i < hashSize; ++i) {
            this.buckets[i] = new List();
        }
        this.hashFunction = hashFunction;
    }

    /**
     * add element to hash table
     *
     * @param string string, which you want to add
     */
    public void addToHash(String string) {
        int index = hashFunction.hashFunction(string, hashSize);
        buckets[index].addToEnd(string);
    }

    /**
     * delete element of the hash table
     *
     * @param string string, which you want to delete
     */
    public void deleteElement(String string) {
        int hash = hashFunction.hashFunction(string, hashSize);
        List.ListElement temp = buckets[hash].searchPosition(string);

        if (temp != null) {
            buckets[hash].deleteElement(temp);
            System.out.println("Successfully deletion");
        } else {
            System.out.println("Element not found");
        }
    }

    /**
     * @return size of hash table
     */
    public int getHashSize() {
        return hashSize;
    }

    public void changeHashFunction(HashFunctionInterface newHashFunction) {
        List[] newBuckets = new List[hashSize];
      
        for (int i = 0; i < hashSize; ++i) {
            newBuckets[i] = new List();
        }

        for (int i = 0; i < hashSize; ++i) {
            List.ListElement position = buckets[i].firstPosition();
            while (position != null) {
                String currentString = buckets[i].positionValue(position);
                int index = newHashFunction.hashFunction(currentString, hashSize);
                newBuckets[index].addToEnd(currentString);
                List.ListElement temp = position;
                position = buckets[i].nextPosition(position);
                buckets[i].deleteElement(temp);
            }        
        }
        
        System.arraycopy(newBuckets, 0, buckets, 0, hashSize);
        this.hashFunction = newHashFunction;
    }

    /**
     * checks existence of the element in the hash table
     *
     * @param string string, which you check for existence
     * @return true if exists; false if not exists
     */
    public boolean Exists(String string) {
        int hash = hashFunction.hashFunction(string, hashSize);
        return buckets[hash].Exist(string);
    }
    /**
     * size of hash table
     */
    final private int hashSize = 100;
    /**
     * hash table elements
     */
    private List[] buckets = new List[hashSize];
    /**
     * hash function
     */
    private HashFunctionInterface hashFunction;
}
